src/code/providers/provider-interface.ts

```diff
@@ -0,0 +1,550 @@
+}
+
+type CloudFileContentType = any
+interface ICloudFileOpts {
+  content: CloudFileContentType
+  metadata: ICloudMetaDataOpts
+}
+
+class CloudFile {
+  content: CloudFileContentType
+  metadata: ICloudMetaDataOpts
+  constructor(options: ICloudFileOpts) {
+    ({content: this.content, metadata: this.metadata} = options)
+  }
+}
+
+class CloudMetadata {
+  name: string
+  description: string
+  content: any
+  url: string
+  type: ICloudFileTypes
+  provider: providerType
+  parent: ICloudMetaDataOpts
+  providerData: providerDataType
+  overwritable: boolean
+  sharedContentId: documentId
+  sharedContentSecretKey: key
+  mimeType: mimeType
+  filename: string|null
+
+  static Folder = 'folder'
+  static File = 'file'
+  static Label = 'label'
+
+  // TODO IMPORTANT: These are defined as class variables,
+  // which seems like a mistake. Seems like it should match
+  // the mimeType, which is an instance variable.
+  static Extension: string|null = null
+  static ReadableExtensions: string[]
+
+  constructor(options: ICloudMetaDataOpts) {
+    let provider, parent, providerData
+    this.name = options.name
+    this.type = options.type
+    this.description = options.description
+    this.content = options.content
+    this.url = options.url
+    provider = options.provider
+    this.provider = provider != null ? provider : null
+    parent = options.parent
+    this.parent = parent
+    providerData = options.providerData
+    this.providerData = providerData != null ? providerData : {}
+    this.overwritable = options.overwritable
+    this.sharedContentId = options.sharedContentId
+    this.sharedContentSecretKey = options.sharedContentSecretKey
+    this.mimeType = options.mimeType
+    this._updateFilename()
+  }
+
+
+  static mapTypeToCloudMetadataType(iType: ICloudFileTypes) {
+    return iType || ICloudFileTypes.File
+  }
+
+  static nameIncludesExtension(name:string) {
+    if (name.indexOf(FILE_EXTENSION_DELIMETER) === -1) {
+      return false
+    }
+    return true
+  }
+
+  static withExtension(name: string, defaultExtension:string , keepOriginalExtension:boolean) {
+    // TODO:  Eslint doesn't like this bitwise not operator.
+    // I feel the same.
+    if (keepOriginalExtension && this.nameIncludesExtension(name)) {
+      return name
+    }
+    const extension = this.Extension || defaultExtension
+    if (extension) {
+      return this.newExtension(name, extension)
+    } else {
+      return name
+    }
+  }
+
+  static newExtension(name: string, extension:string) {
+    // drop last extension, if there is one
+    name = name.substr(0, name.lastIndexOf('.')) || name
+    return name + "." + extension
+  }
+
+  path() {
+    const _path = []
+    let { parent } = this
+    while (parent !== null) {
+      _path.unshift(parent);
+      ({ parent } = parent)
+    }
+    return _path
+  }
+
+  rename(newName: string) {
+    this.name = newName
+    return this._updateFilename()
+  }
+
+  _updateFilename() {
```

Comment:
I would try to make all the underscore methods and variables protected (in this case as it is a base class but private in terminal classes) and see if they are leaking out.